DMD.C

```diff
@@ -0,0 +1,489 @@
+        }
+        double current_energy = 0.0;
+        for (int i = 0; i < num_singular_vectors; i++)
+        {
+            current_energy += d_factorizer->S[i];
+            if (current_energy / total_energy >= d_energy_fraction)
+            {
+                d_k = i + 1;
+                break;
+            }
+        }
+    }
+
+    std::cout << "Using " << d_k << " basis vectors out of " << num_singular_vectors << "." << std::endl;
+
+    // Allocate the appropriate matrices and gather their elements.
+    Matrix* d_basis = new Matrix(f_snapshots->numRows(), d_k, f_snapshots->distributed());
+    Matrix* d_S_inv = new Matrix(d_k, d_k, false);
+
+    Matrix* d_basis_right = new Matrix(f_snapshots_minus->numColumns(), d_k, false);
+    for (int d_rank = 0; d_rank < d_num_procs; ++d_rank) {
+        // V is computed in the transposed order so no reordering necessary.
+        gather_block(&d_basis->item(0, 0), d_factorizer->V,
+                     1, row_offset[static_cast<unsigned>(d_rank)]+1,
+                     d_k, row_offset[static_cast<unsigned>(d_rank) + 1] -
+                     row_offset[static_cast<unsigned>(d_rank)],
+                     d_rank);
+
+        // gather_transposed_block does the same as gather_block, but transposes
+        // it; here, it is used to go from column-major to row-major order.
+        gather_transposed_block(&d_basis_right->item(0, 0), d_factorizer->U, 1, 1,
+                                f_snapshots_minus->numColumns(), d_k, d_rank);
+    }
+
+    // Get inverse of singular values by multiplying by reciprocal.
+    for (int i = 0; i < d_k; ++i)
+    {
+        d_S_inv->item(i, i) = 1 / d_factorizer->S[static_cast<unsigned>(i)];
+    }
+
+    // Calculate A_tilde = U_transpose * f_snapshots_plus * V * inv(S)
+    Matrix* d_basis_mult_f_snapshots_plus = d_basis->transposeMult(f_snapshots_plus);
+    Matrix* d_basis_mult_f_snapshots_plus_mult_d_basis_right = d_basis_mult_f_snapshots_plus->mult(d_basis_right);
+    Matrix* A_tilde = d_basis_mult_f_snapshots_plus_mult_d_basis_right->mult(d_S_inv);
+
+    // Calculate eigenvalues/eigenvectors of A_tilde
+    // Call lapack routines to do the eigensolve.
+    char jobvl, jobrl;
+
+    // Calculate right eigenvectors only.
+    jobvl = 'N';
+    jobrl = 'V';
+
+    int info;
+    int lwork = std::max(d_k*d_k, 10*d_k);
+    double* work = new double [lwork];
+    double* e_real = new double [d_k];
+    double* e_imaginary = new double [d_k];
+    double* ev_l = NULL;
+    Matrix* ev_r = new Matrix(d_k, d_k, false);
+
+    // A_tilde now in a row major representation.  Put it
+    // into column major order.
+    for (int row = 0; row < d_k; ++row) {
+        for (int col = row+1; col < d_k; ++col) {
+            double tmp = A_tilde->item(row, col);
+            A_tilde->item(row, col) = A_tilde->item(col, row);
+            A_tilde->item(col, row) = tmp;
+        }
+    }
+
+    // Now call lapack to do the eigensolve.
```

Comment:
Would it be better to refactor the eigensolve into a new function that could be reused?